Update rocm-core package to include rdhc script

solaiys opened the pull request:
Adds rdhc script in to rocm-core package to install in to <ROCM_INSTALL_PATH>/libexec/rocm-core/.
Creates a symlink for rdhc in bin/ folder.
README.md and requirements.txt files are added in to <ROCM_INSTALL_PATH>share/rdhc/

rdhc.py checks for the required pip packages presence in the system. If its not installed, warns the user and exits gracefully.

solaiys committed:
Add rdhc script in to rocm-core package

Signed-off-by: Saravanan Solaiyappan <saravanan.solaiyappan@amd.com>

solaiys committed:
Add package maintainer scripts for rdhc entries.

Post install script will create a symlink for rdhc in bin/ folder.
prerm script will remove the symlink when it is uninstalled.

Signed-off-by: Saravanan Solaiyappan <saravanan.solaiyappan@amd.com>

nunnikri commented on projects/rocm-core/CMakeLists.txt:
```diff
@@ -149,6 +149,14 @@ if(BUILD_SHARED_LIBS)
 
+  # Install rdhc folder
+  install(DIRECTORY ${CMAKE_CURRENT_SOURCE_DIR}/rdhc/
+          DESTINATION ${CMAKE_INSTALL_DATAROOTDIR}/rdhc
```
If this tool is supposed to have soft link in bin folder, then it should be in libexec folder not share

solaiys replied:
taken care.

arvindcheru commented on projects/rocm-core/CMakeLists.txt:
```diff
@@ -149,6 +149,14 @@ if(BUILD_SHARED_LIBS)
+          DESTINATION ${CMAKE_INSTALL_DATAROOTDIR}/rdhc
+          COMPONENT CORE_RUNTIME
+          FILES_MATCHING
+          PATTERN "*"
```
Request to make explicit file level install rather than folder install

solaiys replied:
taken care.

arvindcheru commented on projects/rocm-core/rdhc/rdhc.py:
```diff
@@ -1748,7 +1748,8 @@ def main():
                                     formatter_class=argparse.RawDescriptionHelpFormatter,
                                     usage="sudo -E ./rdhc.py [options]",
-                                    epilog="Usage examples:\n"+
+                                    epilog="Refer the README @<ROCM_INSTALL_PATH>/share/rdhc/README.md \n"+
```
Can we rename this README.md to have unique name or integrate with Main module level README?
install folder to consider share/doc/rocm-core/?

solaiys committed:
Create the rdhc symlink within the package itself.

postinst and prerm scripts are not needed if the symlink is
created as part of the package itself.

Signed-off-by: Saravanan Solaiyappan <saravanan.solaiyappan@amd.com>

solaiys committed:
[RDHC] Check if the required pip pkgs are present and warn.

rdhc checks the required pip packages are present or not.
if not warns the user and exits gracefully.

Signed-off-by: Saravanan Solaiyappan <saravanan.solaiyappan@amd.com>

solaiys committed:
Updated with the install FILES instead of DIRECTORY.

Removed hard-coding of rocm-core name, used CORE_TARGET instead.

Signed-off-by: Saravanan Solaiyappan <saravanan.solaiyappan@amd.com>

ScottTodd commented on projects/rocm-core/CMakeLists.txt:
I suspect something in this broken downstream Windows builds in TheRock when we tried to integrate this change, see https://github.com/ROCm/TheRock/pull/1824#discussion_r2441435175

```
[75/285] Stage installing sub-project rocm-core
FAILED: base/rocm-core/stamp/stage.stamp B:/build/base/rocm-core/stamp/stage.stamp 
C:\Windows\system32\cmd.exe /C "cd /D B:\build\base\rocm-core\build && C:\home\runner\_work\_tool\Python\3.12.10\x64\python3.exe C:/home/runner/_work/TheRock/TheRock/build_tools/teatime.py --log-timestamps --label "rocm-core install" --no-interactive B:/build/logs/rocm-core_install.log -- "C:/Program Files/CMake/bin/cmake.exe" --install B:/build/base/rocm-core/build && C:\home\runner\_work\_tool\Python\3.12.10\x64\python3.exe C:/home/runner/_work/TheRock/TheRock/build_tools/fileset_tool.py copy --verbose B:/build/base/rocm-core/dist B:/build/base/rocm-core/stage && "C:\Program Files\CMake\bin\cmake.exe" -E touch B:/build/base/rocm-core/stamp/stage.stamp"
rmtree B:\build\base\rocm-core\dist
mkdir B:\build\base\rocm-core\dist\.info
hardlink B:\build\base\rocm-core\stage\.info\version -> B:\build\base\rocm-core\dist\.info\version
mkdir B:\build\base\rocm-core\dist\bin

Traceback (most recent call last):
  File "C:\home\runner\_work\TheRock\TheRock\build_tools\fileset_tool.py", line 237, in <module>
    main(sys.argv[1:])
  File "C:\home\runner\_work\TheRock\TheRock\build_tools\fileset_tool.py", line 233, in main
    args.func(args)
  File "C:\home\runner\_work\TheRock\TheRock\build_tools\fileset_tool.py", line 138, in run_action
    action(args, pm)
  File "C:\home\runner\_work\TheRock\TheRock\build_tools\fileset_tool.py", line 37, in do_copy
    pm.copy_to(
  File "C:\home\runner\_work\TheRock\TheRock\build_tools\_therock_utils\pattern_match.py", line 146, in copy_to

    if direntry.is_dir() and not direntry.is_symlink():

       ^^^^^^^^^^^^^^^^^

OSError: [WinError 123] The filename, directory name, or volume label syntax is incorrect: 'B:\\build\\base\\rocm-core\\stage\\bin\\rdhc'
```

Logs: https://github.com/ROCm/TheRock/actions/runs/18602498717/job/53044149585?pr=1824#step:13:168

Maybe treating that as a directory since it has no file extension? I haven't seen that style of error before, so this is doing something different than other projects on Linux+Windows.

solaiys replied:
I followed the steps from the projects/rocm-smi-lib/rocm_smi/CMakeLists.txt for the symlink rocm-cmi.
Though there is slight difference w.r.t PERMISSIONS steps for the symlink, it work fine for linux (deb & rpm).
Let me change the code same as the implementation done for the rocm-cmi symlink and submit a PR..

Meanwhile, 
Please see if the change is required in the <TheRock\build_tools\_therock_utils\pattern_match.py> for the robust handling of symlink name & dir name.
May be reordering the checks and avoid following symlinks might prevent this [ WinError 123]
>>>> if direntry.is_dir() and not direntry.is_symlink():

ScottTodd replied:
Let's revert.

Symlink support is not universally available on Windows. and I'm not seeing any other uses of symlinks across the build system that aren't also gated by `if(NOT WIN32)` or `if(CMAKE_SYSTEM_NAME STREQUAL "Linux")`.

If the tool really belongs in the install `bin/` folder, can it be copied there instead, or can we find some other more robust and portable way to distribute it?

ScottTodd replied:
Revert sent for review: https://github.com/ROCm/rocm-systems/pull/1464

lamikr commented:
This is the command causing privilege error on Windows when this patch try to create a symlink:

```
[rocm-core] C:\WINDOWS\system32\cmd.exe /C "cd /D C:\rock\therock\build\base\rocm-core\build && C:\rock\therock\.venv\Lib\site-packages\cmake\data\bin\cmake.exe -E create_symlink ../libexec/rocm-core/rdhc.py rdhc-link-static"
[rocm-core] CMake Error: failed to create symbolic link 'rdhc-link-static': A required privilege is not held by the client.

```
